src/handlers/egress-tracker.js

```diff
@@ -0,0 +1,117 @@
+    }
+
+    if (!response.ok || !response.body) {
+      return response
+    }
+
+    const { dataCid } = ctx
+    const accounting = ctx.ACCOUNTING_SERVICE ?? Accounting.create({
+      serviceURL: env.ACCOUNTING_SERVICE_URL
+    })
+
+    const { readable, writable } = createEgressPassThroughStream(ctx, accounting, dataCid)
+
+    try {
+      ctx.waitUntil(response.body.pipeTo(writable))
+    } catch (error) {
+      console.error('Error in egress tracker handler:', error)
+      // Original response in case of an error to avoid breaking the chain and serve the content
+      return response
+    }
+
+    return new Response(readable, {
+      status: response.status,
+      statusText: response.statusText,
+      headers: response.headers
+    })
+  }
+}
+
+/**
+ * Creates a TransformStream to count bytes served to the client.
+ * It records egress when the stream is finalized without an error.
+ *
+ * @param {import('@web3-storage/gateway-lib/middleware').Context} ctx - The context object.
+ * @param {AccountingService} accounting - The accounting service instance to record egress.
+ * @param {import('@web3-storage/gateway-lib/handlers').CID} dataCid - The CID of the served content.
+ * @returns {TransformStream} - The created TransformStream.
+ */
+function createEgressPassThroughStream (ctx, accounting, dataCid) {
+  let totalBytesServed = 0
+
+  return new TransformStream({
+    /**
+     * The start function is called when the stream is being initialized.
+     * It resets the total bytes served to 0.
+     */
+    start () {
+      totalBytesServed = 0
+    },
+    /**
+     * The transform function is called for each chunk of the response body.
+     * It enqueues the chunk and updates the total bytes served.
+     * If an error occurs, it signals an error to the controller and logs it.
+     * The bytes are not counted in case of enqueuing an error.
+     * @param {Uint8Array} chunk
+     * @param {TransformStreamDefaultController} controller
+     */
+    async transform (chunk, controller) {
+      try {
+        controller.enqueue(chunk)
+        totalBytesServed += chunk.byteLength
+      } catch (error) {
+        console.error('Error while counting egress bytes:', error)
+        controller.error(error)
+      }
+    },
+
+    /**
+     * The flush function is called when the stream is being finalized,
+     * which is when the response is being sent to the client.
+     * So before the response is sent, we record the egress.
+     * It is called only once and it triggers a non-blocking call to the accounting service.
+     * If an error occurs, the egress is not recorded.
+     * NOTE: The flush function is NOT called in case of an stream error.
+     */
+    async flush (controller) {
+      try {
+        // Non-blocking call to the accounting service to record egress
+        if (totalBytesServed > 0) {
+          ctx.waitUntil(accounting.record(dataCid, totalBytesServed, new Date().toISOString()))
+        }
```

Comment:
**thought:** I notice that:

* `dataCid` and `accounting` are arguments that we were passed.
* `ctx.waitUntil` is the only thing we use from the `ctx`.
* Since anyone can `new Date().toISOString()`, that leaves `totalBytesServed` (understandably) as the only thing that this function is really responsible for creating.
* It's a little odd that `createEgressPassThroughStream` knows how to communicate with the Accounting Service.

**suggestion:** What about something like:

```ts
/**
 * @param {(totalBytesServed: number) => void} onClose
 * @returns {TransformStream}
 */
function createByteCountStream(onClose) {
  let totalBytesServed = 0;

  return new TransformStream({
    // ...
    async flush() {
      onClose(totalBytesServed);
    },
  });
}

const responseBody = response.body.pipeThrough(
  createByteCountStream((totalBytesServed) => {
    // Non-blocking call to the accounting service to record egress
    if (totalBytesServed > 0) {
      ctx.waitUntil(
        accounting.record(dataCid, totalBytesServed, new Date().toISOString())
      );
    }
  })
);
```

Reply:
reasonable

Reply:
will certainly simplify the test